adapters/coap-vertx-base/src/main/java/org/eclipse/hono/adapter/coap/CoapCredentialsStore.java

```diff
@@ -0,0 +1,268 @@
+                }
+                secret.complete(key);
+            }, secret);
+        }
+    }
+
+    /**
+     * Create tenant aware identity based on the provided pre-shared-key handshake identity.
+     * 
+     * @param identity pre-shared-key handshake identity.
+     * @return tenant aware identity.
+     */
+    public PreSharedKeyTenantIdentity getHandshakeIdentity(String identity) {
+        String splitRegex = config.isSingleTenant() ? null : config.getIdSplitRegex();
+        return PreSharedKeyTenantIdentity.create(identity, splitRegex);
+    }
+
+    @Override
+    public byte[] getKey(String identity) {
+        LOG.info("get secret key for {}", identity);
+        PreSharedKeyTenantIdentity handshakeIdentity = getHandshakeIdentity(identity);
+        if (handshakeIdentity == null) {
+            return null;
+        }
+        //
+        CompletableFuture<byte[]> secret = new CompletableFuture<>();
+        vertx.runOnContext((v) -> {
+            getSharedKeyForDevice(handshakeIdentity).setHandler((result) -> secret.complete(result.result()));
+        });
+        try {
+            // timeout, don't block handshake too long
+            return secret.get(10000, TimeUnit.MILLISECONDS);
+        } catch (InterruptedException e) {
+        } catch (ExecutionException e) {
+        } catch (TimeoutException e) {
+        }
+        LOG.warn("missing secret key for {}!", identity);
+        return null;
+    }
+
+    @Override
+    public byte[] getKey(ServerNames serverNames, String identity) {
+        // for now, don't support serverNames indication
+        // maybe extended in the future to provide tenant
+        // would require solution for californium issue #511
+        return getKey(identity);
+    }
+
+    @Override
+    public String getIdentity(InetSocketAddress inetAddress) {
+        // not used by dtls server, and role exchange not supported!
+        return null;
+    }
+
+    /**
+     * Get hono device for pre-shared-key handshake identity from cache.
+     * 
+     * @param identity identity provided by pre-shared-key handshake. Contains tenant and authentication id.
+     * @return hono device, or {@code null}, if not available in cache.
+     */
+    public Device getCachedDevice(String identity) {
+        PreSharedKeyTenantIdentity handshakeIdentity = getHandshakeIdentity(identity);
+        if (handshakeIdentity != null) {
+            return devices.getIfPresent(handshakeIdentity);
+        }
+        return null;
+    }
+
+    /**
+     * Get authorized hono device from URI identity.
+     * 
+     * Uses {@link #credentialsProvider}'s
+     * {@link CredentialsApiAuthProvider#authenticate(DeviceCredentials, io.vertx.core.Handler)} to authenticate the
+     * identity extracted from the provided URI.
+     * 
+     * @param exchange coap exchange with URI.
+     * @param authLevel path level to extract URI identity. e.g. for "/telemetry/tenant/authid/passbase64" the level is
```

Comment:
I haven't seen this authentication scheme before. Where does this come from? My understanding is that CoAP clients are authenticated as part of the DTLS handshake or am I mistaken?

Reply:
It's inspired by similar approaches as the "gateway mode". Maybe it never goes into hono, if certificates could be handle in the credentials store.

Reply:
Hi @sophokles73 

I will check, if I can use the credentials service for client certificates. If that works, I remove this.